packages/dds/tree/src/feature-libraries/modular-schema/fieldKind.ts

```diff
@@ -0,0 +1,161 @@
+     * listing the other old kind here can prevent those edits from being conflicted and
+     * provide a chance to handle them.
+     */
+    public constructor(
+        public readonly identifier: FieldKindIdentifier,
+        public readonly multiplicity: Multiplicity,
+        public readonly changeHandler: ChangeHandler<any>,
+        private readonly allowsTreeSupersetOf:
+            (originalTypes: ReadonlySet<TreeSchemaIdentifier> | undefined, superset: FieldSchema) => boolean,
+        public readonly handlesEditsFrom: ReadonlySet<FieldKindIdentifier>,
+        ) {}
+
+    /**
+     * @returns true iff `superset` permits a (non-strict) superset of the subtrees
+     * allowed by field made from `this` and `originalTypes`.
+     */
+    public allowsFieldSuperset(
+        policy: FullSchemaPolicy,
+        originalData: SchemaData,
+        originalTypes: ReadonlySet<TreeSchemaIdentifier> | undefined,
+        superset: FieldSchema,
+    ): boolean {
+        if (isNeverField(policy, originalData, fieldSchema(this, originalTypes))) {
+            return true;
+        }
+        if (isNeverField(policy, originalData, superset)) {
+            return false;
+        }
+        return this.allowsTreeSupersetOf(originalTypes, superset);
+    }
+}
+
+/**
+ * Policy from the app for interpreting the stored schema.
+ * The app must ensure consistency for all users of the document.
+ */
+ export interface FullSchemaPolicy extends SchemaPolicy {
+    /**
+     * Policy information about FieldKinds:
+     * This is typically stored as code, not in documents, and defines how to handles fields based on their kind.
+     * It is assumed that all users of a document will have exactly the same FieldKind policies,
+     * though older applications might be missing some,
+     * and will be unable to process any changes that use those FieldKinds.
+     */
+    readonly fieldKinds: ReadonlyMap<FieldKindIdentifier, FieldKind>;
+}
+
+/**
+ * Functionality provided by a field kind which will be composed together to
+ * implement a unified ChangeFamily supporting documents with multiple field kinds.
+ *
+ * TODO: eventually field-kinds will need to provide everything ChangeFamily requires.
+ */
+export interface ChangeHandler<TChange> {
+    /**
+     * Provides merge / rebase policy for handling concurrent changes.
+     *
+     * TODO: This might handle changes from other field kinds if the file kind opts into them.
+     * Sort out how to reflect this in the type parameter for this vs other uses of TChanges in this interface.
+     */
+    readonly rebaser: ChangeRebaser<TChange>;
+
+    /**
+     * Handles encoding of changes.
+     */
+    readonly encoder: ChangeEncoder<TChange>;
+
+    // TODO: add edit builder.
+}
+
+/**
+ * Describes how a particular field functions.
+ *
+ * This determine its reading and editing APIs, multiplicity, and what merge resolution policies it will use.
+ */
+export enum Multiplicity {
```

Comment:
It's not a use-case we have run into or plan to support in the near term. It could always be added later or if the TODO above is done to replace this with numeric bounds it would also be expressible.

Reply:
Note that this type was just moved, and remains unchanged. If we want to change it I don't think it should be in this PR.

Reply:
Ah cool. Yeah, if there's no immediate use case, I don't think there's a point in adding it.